pkg/watch/manager.go

```diff
@@ -223,26 +223,32 @@ func (wm *Manager) doRemoveWatch(r *Registrar, gvk schema.GroupVersionKind) erro
 	return nil
 }
 
-// replaceWatches ensures all and only desired watches are running.
-func (wm *Manager) replaceWatches(r *Registrar) error {
+// reconcileWatches ensures all and only desired watches are running.
+func (wm *Manager) reconcileWatches() error {
 	wm.watchedMux.Lock()
 	defer wm.watchedMux.Unlock()
 
 	var errlist errorList
 
 	desired := wm.managedKinds.Get()
-	for gvk := range wm.watchedKinds {
-		if _, ok := desired[gvk]; !ok {
+	for gvk, v := range wm.watchedKinds {
```

Comment:
I think I see what you're saying about how two registrar's watches can bleed into each other. It looks like the problem here is that if you have r1 and r2 both watching GVK-A, if r2 wants to remove GVK-A via a `replaceWatches()` invocation, it is blocked because r1 is keeping GVK-A in `wm.managedKinds.Get()`, triggering the short-circuit check on `desired`.

It looks like this change doesn't address that problem. I'm wondering if the solution is to just drop the short-circuit and call `doRemoveWatch()` for any watched GVK not in the registrar's roster, relying on the short-circuit in `doRemoveWatch()` to handle the case where the registrar never caused a watch on the resource in the first place.

Reply:
Good call - I'm adding a test case to capture that as well.